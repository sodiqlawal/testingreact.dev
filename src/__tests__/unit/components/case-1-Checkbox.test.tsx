import React from 'react'
// import ReactDOM from 'react-dom'
import { axe } from 'jest-axe'
import { render, fireEvent } from '@testing-library/react'

import Checkbox, { CheckboxPropsInterface } from '../../../components/Checkbox'

/**
 * This checkbox component renders a checkbox with a label.
 * Since we customized the default checkbox, we want to
 * make sure it still works as a regular checkbox
 * should.
 */
describe('The <Checkbox /> component', () => {
  const defaultCheckProps: CheckboxPropsInterface = {
    label: 'TEST_CHECKBOX_LABEL',
    id: 'TEST_CHECKBOX_ID',
    background: '#000',
    checkMarkBackground: '#fff',
    onChange: jest.fn(),
    checked: false,
  }
  const setupCheckbox = (props = defaultCheckProps) =>
    render(<Checkbox {...props} />)

  it('Should render the label and checkbox the user will see', () => {
    // using react dom
    // const container = document.createElement('div')
    // ReactDOM.render(<Checkbox {...defaultCheckProps} />, container)
    // expect(container.querySelector('label')).not.toBeNull()
    // expect(container.querySelector('input[type="checkbox"]')).not.toBeNull()

    // using react testing library
    const { asFragment } = setupCheckbox()

    expect(asFragment()).toMatchSnapshot()
    // debug(getByLabelText('TEST_CHECKBOX_LABEL'))
    // expect(getByLabelText('TEST_CHECKBOX_LABEL')).toBeInTheDocument()
  })

  it('Should make the checkbox accessible by setting the id and htmlFor attributes on label and checkbox', () => {
    const { getByLabelText } = setupCheckbox()
    expect(getByLabelText(defaultCheckProps.label)).toBeInTheDocument()
  })

  it('Should call the onChange handler when it is provided', () => {
    const { getByLabelText } = setupCheckbox()
    const checkbox = getByLabelText(defaultCheckProps.label)
    fireEvent.click(checkbox)
    expect(defaultCheckProps.onChange).toHaveBeenCalled()
  })

  it('Should change state correctly when clicked (checked and unchecked)', () => {
    const { getByLabelText } = setupCheckbox({
      ...defaultCheckProps,
      checked: true,
    })

    expect(getByLabelText(defaultCheckProps.label)).toBeChecked()
  })

  it('should not fail any accessibility tests', async () => {
    const { container } = setupCheckbox()
    expect(await axe(container)).toHaveNoViolations()
  })
})
