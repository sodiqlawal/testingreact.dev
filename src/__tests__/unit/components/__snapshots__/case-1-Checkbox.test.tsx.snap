// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The <Checkbox /> component Should render the label and checkbox the user will see 1`] = `
<DocumentFragment>
  <div
    class="sc-bdfBwQ"
    data-testid="Checkbox"
  >
    <input
      class="sc-hKgILt btwhit"
      id="TEST_CHECKBOX_ID"
      type="checkbox"
    />
    <label
      class="sc-dlfnbm fGbBqX"
      for="TEST_CHECKBOX_ID"
    >
      <span
        class="sc-gsTCUz eOcHst"
      >
        TEST_CHECKBOX_LABEL
      </span>
    </label>
  </div>
</DocumentFragment>
`;
